### ISSUE TITLE
Optional as return type seems to be broken (JDK 8 + 9)

### ISSUE THREAD

**Issue by ThomasGP** [SvenEwald/xmlbeam#46]
I am using xmlbeam 1.4.13 with the newest JDK 9.0.1. In my projection interface I have a method with return type `Optional<String>`. If I call this method, an `IllegalArgumentException` is raised (from [line 362](https://github.com/SvenEwald/xmlbeam/blob/edff94e24c24bf6c446f2c42292e709d94f8416c/src/main/java/org/xmlbeam/ProjectionInvocationHandler.java#L362) in `ProjectionInvocationHandler.ReadInvocationHandler`).

This is due to the control flow in that class: With the `DefaultTypeConverter` in use, an Optional return type is deemed unconvertable (`isConvertable == false`). Consequentially, lines 324-326 will not get executed, the lines that are supposed to deal with an Optional.

My fix for now is to add an Optional conversion to the type converter:
```java
private static void fixOptionalTypeConversion(XBProjector projector) {
  DefaultTypeConverter typeConverter = projector.config().getTypeConverterAs(DefaultTypeConverter.class);
  typeConverter.setConversionForType(Optional.class, new DefaultTypeConverter.Conversion<>(null) {
    @Override
    public Optional convert(String data) {
      return Optional.ofNullable(data);
    }
  });
}
```

With that fix, the projection interface method returns an Optional as expected.

Can you confirm this bug? I noticed the [test repository for Java 8](https://github.com/SvenEwald/xbjava8integration) having some [tests](https://github.com/SvenEwald/xbjava8integration/blob/ed50ae45a267289c1663055ec389be2159d138cc/src/test/java/org/xmlbeam/tests/java8/TestOptionalAsReturnType.java#L11) for the Optional return type that pretty much resemble my situation.

These tests fail if I run them against xmlbeam 1.4.13 (with JDKs 9.0.1 and 1.8.0_152). Are they still relevant and maintained? I noticed in the pom.xml that the project depends on xmlbeams 1.4.7-SNAPSHOT.

**Comment by SvenEwald** [comment]
Hi Thomas, thank you very much for your bug report. I can not yet confirm the bug, I have to try with newer JDK versions. It is working on 1.8.0-b132 (MacOs) and 1.8.0_131-b11 (ARM), but I have an Idea why this could break in newer JDKs. 
Your fix looks like a good workaround until I have found a fix.

**Comment by SvenEwald** [comment]
It's working with 1.8.0_102 (MacOs).

**Comment by ThomasGP** [comment]
To further illustrate: I created a [repo](https://github.com/ThomasGP/xmlbeam-issue) with the Optional tests, just as they are in your Java 8 repo. If I run `mvn clean verify` in the root folder, the build fails due to tests in error:

> testValueIsNotPresent(MainTest): Return type class java.lang.Integer of method public abstract java.util.Optional MainTest$OptionalReturningProjection.getSomeValue() is not supported. Please change to an projection interface, a List, an Array or one of current type converters types:org.xmlbeam.types.DefaultTypeConverter@17ed40e0

> testValueIsPresent(MainTest): Return type class java.lang.Integer of method public abstract java.util.Optional MainTest$OptionalReturningProjection.getSomeValue() is not supported. Please change to an projection interface, a List, an Array or one of current type converters types:org.xmlbeam.types.DefaultTypeConverter@71e7a66b

According to `mvn -version`, the maven build runs with:

> Apache Maven 3.5.2 (138edd61fd100ec658bfa2d307c43b76940a5d7d; 2017-10-18T09:58:13+02:00)
Maven home: D:\Maven\apache-maven-3.5.2
Java version: 1.8.0_152, vendor: Oracle Corporation
Java home: C:\Program Files\Java\jdk1.8.0_152\jre
Default locale: en_US, platform encoding: Cp1252
OS name: "windows 7", version: "6.1", arch: "amd64", family: "windows"

If I switch to JDK 9 (changing the versions in the maven-compiler-plugin config inside the pom.xml, too), the same thing happens. `mvn -version` in that case is:

> Apache Maven 3.5.2 (138edd61fd100ec658bfa2d307c43b76940a5d7d; 2017-10-18T09:58:13+02:00)
Maven home: D:\Maven\apache-maven-3.5.2
Java version: 9.0.1, vendor: Oracle Corporation
Java home: C:\Program Files\Java\jdk-9.0.1
Default locale: en_US, platform encoding: Cp1252
OS name: "windows 7", version: "6.1", arch: "amd64", family: "windows"

Please let me know how to give further information to reproduce this behavior.

**Comment by SvenEwald** [comment]
Excellent! I'll check out your repo. I could not reproduce the error with 1.8.0_152 (Mac)...

**Comment by SvenEwald** [comment]
Now I can confirm the error. Thank you very much!

**Comment by SvenEwald** [comment]
First  error found in java8 integration repo. The test dependency was pinned to XMLBeam 1.4.7. This is the reason why I did not see this error. I think I know what causes the error. It is not related to the Java version, but to a refactoring I did after 1.4.7.

**Comment by SvenEwald** [comment]
Fixed in 1.4.14 (coming soon).
